### ISSUE TITLE
Add BlocSparta as dependecy of BlocMorphic?

### ISSUE THREAD

**Issue by tinchodias** [pharo-graphics/Bloc#81]
...or maybe only SpartaCore?

History:

In pharo/bloc channel, Aliaksei said: "...to be able to render bloc inside of the morphic we can load morphic host (independent from Sparta)" and recommended loading BlocMorphic.

When I load:
~~~
Metacello new
    baseline: 'BlocMorphic';
    repository: 'github://pharo-graphics/Bloc/src';
    load.
~~~

And execute this suggested snippet:
~~~
element := BlElement new
    size: 200@100;
    background: Color blue.
space := BlSpace new.
space title: 'Bloc Window'.
space extent: 400@200.
space addChild: element.
space show.
~~~

I get `Instance of BlMorphicWindowHostSpace did not understand #withSpartaRenderer`

Which doesn't happen if I load "BlocSparta".

**Comment by syrel** [comment]
Hi @tinchodias 

`Bloc Host` **are** independent from the renderer. it took us very very long time and a few rewrites to reach this point. However, because we just have one renderer we never got to implement a renderer selection mechanism (could be based on subclasses or pragmas).

Basically, instead of `withSpartaRenderer` it should send something like `withBestRenderer`.
If we look at the implemention:
```
withSpartaRenderer
	renderer := BlSpartaRenderer new
```
it is clear that renderer does not depend on anything. it is just a simple object that could be easily injected from the outside

It would be great to preserve and enforce this property.

**Comment by tinchodias** [comment]
@syrel  
Thank you very much for the explanation. I'm learning about Bloc, Sparta, and related.
I still have to grasp some key concepts of these projects. The difference between "Bloc host" and "Bloc renderer" is some of them.
In this issue, I reported my observation in kind of "thinking aloud" way.

**Comment by tinchodias** [comment]
The class comment of `BlMorphicWindowHostSpace` ends with "Depends on Bloc-Sparta". 
(The class is categorized as 'BlocHost-Morphic-Host - Window'.)

IIUC, this sentence is not true, right?

**Comment by tinchodias** [comment]
Hello again @syrel, I analyzed the idea of `withBestRenderer`. You said:
- "Bloc Host are independent from the renderer"
- "implement a renderer selection mechanism (could be based on subclasses or pragmas)."... this would be in the new `withBestRenderer`.

To understand more, I opened a GT image, loaded BlocMorphic, and then took 2 facts:

**Fact 1**: The only implementor of `withSpartaRenderer` is the trait named `TBlHostSpace`, and there are several concrete users.

Concretely, `TBlHostSpace users` answers a collection with `BlGlutinHostSpace BlHeadlessHostSpace BlMorphicWindowHostSpace BlMockedHostSpace BlMorphicHostSpace`

**Fact 2**: The senders of `withSpartaRenderer` are subclasses of `BlHost`.

Concretely, `BlHost allSubclasses` answers a collection with `BlHeadlessHost BlMockedHost BlMorphicSteppingHost BlGlutinHost BlMorphicHost BlMorphicWindowHost`.

Now... **what's "best"?** when there is only one option and it works, like right now when BlocSparta is loaded then okay, but what's best if there is also a new `BlMorphicRenderer`?

First, I thought on creating this new class `BlMorphicRenderer` and adding:
~~~smalltalk
BlMorphicWindowHostSpace >>
withBestRenderer
    renderer := BlMorphicRenderer new
~~~
and same for `BlMorphicHostSpace`. But then there is no independence between host and renderer. 

Second try. You said "based on subclasses", then I imagine `withBestRenderer` delegating the selection to a new method `BlHostRenderer class >> bestRenderer` that sorts its subclasses, sorts them by `priority`, and selects the highest. But sure if there a priority between renderers, anyway.

Third try. There is some additional condition that the renderers can have to work with a host (I don't know this). Then, we add `BlHostRenderer class >> bestRendererFor: aHostSpace` that is similar than in the previous try but filters the subclasses ofo renderer with a `canRenderOn: aHostSpace`.

Thank you in advance for your answer.
